docs/source/basics/messages.rst

```diff
@@ -0,0 +1,144 @@
+    msg = Message.from_dict(
+        {
+            "role": "user",
+            "content": "Hello world!",
+            "masked": True,
+            "eot": True,
+            "ipython": False,
+        },
+    )
+    print(msg.content)
+    # [{'type': 'text', 'content': 'Hello world!'}]
+
+Content is formatted as a list of dictionaries. This is because Messages can also contain multimodal content, such as images.
+
+Images in Messages
+^^^^^^^^^^^^^^^^^^
+For multimodal datasets, you need to add the image as a :class:`~PIL.Image.Image` to the corresponding :class:`~torchtune.data.Message`.
+To add it to the beginning of the message, simply prepend it to the content list.
+
+.. code-block:: python
+
+    import PIL
+    from torchtune.data import Message
+
+    img_msg = Message(
+        role="user",
+        content=[
+            {
+                "type": "image",
+                # Place your image here
+                "content": PIL.Image.new(mode="RGB", size=(4, 4)),
+            },
+            {"type": "text", "content": "What's in this image?"},
+        ],
+    )
+
+This will indicate to the model tokenizers where to add the image special token and will be processed by the model transform
+appropriately.
+
+In many cases, you will have an image path instead of a raw :class:`~PIL.Image.Image`. You can use the :func:`~torchtune.data.load_image`
+utility for both local paths and remote paths.
+
+.. code-block:: python
+
+    import PIL
+    from torchtune.data import Message, load_image
+
+    image_path = "path/to/image.jpg"
+    img_msg = Message(
+        role="user",
+        content=[
+            {
+                "type": "image",
+                # Place your image here
+                "content": load_image(image_path),
+            },
+            {"type": "text", "content": "What's in this image?"},
+        ],
+    )
+
+If your dataset contain image tags, or placeholder text to indicate where in the text the image should be inserted,
+you can use the :func:`~torchtune.data.format_content_with_images` to split the text into the correct content list
+that you can pass into the content field of Message.
+
+.. code-block:: python
+
+    import PIL
+    from torchtune.data import format_content_with_images
+
+    content = format_content_with_images(
+        "<|image|>hello <|image|>world",
+        image_tag="<|image|>",
+        images=[PIL.Image.new(mode="RGB", size=(4, 4)), PIL.Image.new(mode="RGB", size=(4, 4))]
+    )
+    print(content)
+    # [
+    #     {"type": "image", "content": <PIL.Image.Image>},
+    #     {"type": "text", "content": "hello "},
+    #     {"type": "image", "content": <PIL.Image.Image>},
+    #     {"type": "text", "content": "world"}
+    # ]
+
+Message transforms
+^^^^^^^^^^^^^^^^^^
+Message transforms are convenient utilities to format raw data into a list of torchtune :class:`~torchtune.data.Message`
+objects. See :ref:`message_transform_usage_label` for more discussion.
```

Comment:
Message transforms and "Formatting messages with prompt templates" feels a bit out of place. Maybe it should be part of the previous sections